### PR TITLE
Feature/diff pagination

### PR DESCRIPTION
**Pull request by nopcoder**
- API diff support pagination
- API merge return commit reference and summary (counters)
- webui: objects list uncommitted without diff
- webui: changes tab to list uncommitted changes
- webui: compare page should list changes
- webui: revert single change

### CODE REVIEW

File: catalog/cataloger_diff.go [hunk 1]
```diff
@@ -10,17 +10,30 @@ import (
...
 		{Name: "repository", IsValid: ValidateRepositoryName(repository)},
 		{Name: "leftBranch", IsValid: ValidateBranchName(leftBranch)},
 		{Name: "rightBranch", IsValid: ValidateBranchName(rightBranch)},
 	}); err != nil {
-		return nil, err
+		return nil, false, err
 	}
-	differences, err := c.db.Transact(func(tx db.Tx) (interface{}, error) {
+
+	if limit <= 0 || limit > DiffMaxLimit {
```
Comment (tzahij) [inline catalog/cataloger_diff.go:33]: a bit confused. I thought diff does not get limit

File: catalog/cataloger_diff.go [hunk 2]
```diff
@@ -59,56 +78,79 @@ func (c *cataloger) doDiffByRelation(tx db.Tx, relation RelationType, leftID, ri
...
 	}
-	return diffReadDifferences(tx)
+	return nil
 }
 
-func diffReadDifferences(tx db.Tx) (Differences, error) {
+func getDiffDifferences(tx db.Tx, limit int, after string) (Differences, error) {
 	var result Differences
-	if err := tx.Select(&result, "SELECT diff_type, path FROM "+diffResultsTableName); err != nil {
+	query, args, err := psql.Select("diff_type", "path").
```
Comment (tzahij) [inline catalog/cataloger_diff.go:137]: HOPE TO HAVE THIS WORKING SOON, SO i A NOT SURE IT IS NEEDED

Comment (tzahij) [reply catalog/cataloger_diff.go:137]: I take Thursday off, So I guess it will be ready by mid next week

File: catalog/cataloger_merge.go [hunk 3]
```diff
@@ -58,16 +64,16 @@ func (c *cataloger) Merge(ctx context.Context, repository, leftBranch, rightBran
 		}
 
 		if message == "" {
-			message = formatMergeMessage(leftBranch, rightBranch)
+			message = formatMergeMessage(sourceBranch, destinationBranch)
 		}
 		commitID, err := c.doMergeByRelation(tx, relation, leftID, rightID, committer, message, metadata)
```
Comment (tzahij) [inline catalog/cataloger_merge.go:69]: If branches cahnged to source and destination, ID should as well

File: nessie/sanity_api_test.go [hunk 4]
```diff
@@ -121,7 +121,8 @@ func TestSanityAPI(t *testing.T) {
 	diffResp, err = client.Refs.DiffRefs(refs.NewDiffRefsParamsWithContext(ctx).
 		WithRepository(repo).
 		WithLeftRef("branch1").
-		WithRightRef(masterBranch), nil)
+		WithRightRef(masterBranch).
```
Comment (tzahij) [inline nessie/sanity_api_test.go:124]: Sorry, have nothing to say about last three

File: catalog/cataloger_diff.go (hunk 1 above)
Comment (nopcoder) [reply catalog/cataloger_diff.go:33]: This is the public `Diff` function - it should get a limit for pagination.
The actual diff function `doDiff` doesn't get any limit, we limit the part we read from the temporary table.

File: catalog/cataloger_diff.go (hunk 2 above)
Comment (nopcoder) [reply catalog/cataloger_diff.go:137]: Sure. Also there is DiffUncommitted - but I think the current solution should (unless you found an issue there)

File: catalog/cataloger_merge.go (hunk 3 above)
Comment (nopcoder) [reply catalog/cataloger_merge.go:69]: I see the problem - the Cataloger interface defined the method with source and destination when the implementation used left and right. When I update the interface signature, it also updates the parameter names.

File: nessie/sanity_api_test.go (hunk 4 above)
Comment (nopcoder) [reply nessie/sanity_api_test.go:124]: Good - there are the changes needed to make the new API changes work

**Comment by nopcoder** [comment]
updated the PR with webui code changes

File: swagger.yml [hunk 5]
```diff
@@ -66,17 +66,24 @@ definitions:
...
+  merge_result_summary_item:
     type: object
     properties:
       type:
         type: string
         enum: [added, removed, changed, conflict]
-      path:
-        type: string
-      path_type:
+      count:
```
Comment (nopcoder) [inline swagger.yml:75]: have the enum value as object properties

File: catalog/cataloger_diff.go [hunk 6]
```diff
@@ -10,17 +10,30 @@ import (
 	"github.com/treeverse/lakefs/logging"
 )
 
-const diffResultsTableName = "catalog_diff_results"
+type DiffTypeCount struct {
```
Comment (nopcoder) [inline catalog/cataloger_diff.go:13]: model.go + private?

File: catalog/cataloger_diff.go (hunk 1 above)
Comment (nopcoder) [inline catalog/cataloger_diff.go:33]: limit 0 should not set max limit - return nothing but with hasMore value

File: catalog/cataloger_diff_uncommitted.go [hunk 7]
```diff
@@ -9,14 +9,18 @@ import (
...
 	if err := Validate(ValidateFields{
 		{Name: "repository", IsValid: ValidateRepositoryName(repository)},
 		{Name: "branch", IsValid: ValidateBranchName(branch)},
 	}); err != nil {
-		return nil, err
+		return nil, false, err
 	}
-	differences, err := c.db.Transact(func(tx db.Tx) (interface{}, error) {
+
+	if limit <= 0 || limit > DiffMaxLimit {
```
Comment (nopcoder) [inline catalog/cataloger_diff_uncommitted.go:20]: handle limit 0 to return hasMore without results

File: catalog/cataloger_diff_test.go [hunk 8]
```diff
@@ -5,10 +5,78 @@ import (
...
+			c.DeleteEntry(ctx, repository, "branch1", "/file"+strconv.Itoa(i)))
+	}
+	_, err := c.Commit(ctx, repository, "branch1", "delete some files", "tester", nil)
+	testutil.MustDo(t, "delete some files from branch1", err)
+
+	const limit = 3
+	var after string
+	var differences Differences
+	for {
+		res, hasMore, err := c.Diff(ctx, repository, "branch1", "master", limit, after)
```
Comment (nopcoder) [inline catalog/cataloger_diff_test.go:39]: verify that page size eq. limit

File: catalog/cataloger_merge_test.go [hunk 9]
```diff
@@ -113,22 +121,29 @@ func TestCataloger_Merge_FromParentConflicts(t *testing.T) {
...
+	if res.Reference != "" {
+		t.Errorf("Merge reference = %s, expected to be empty", res.Reference)
+	}
+	// TODO(barak): enable test after diff between commits is supported
+	//commitLog, err := c.GetCommit(ctx, repository, reference)
+	//testutil.MustDo(t, "get merge commit reference", err)
+	//if len(commitLog.Parents) != 2 {
+	//	t.Fatal("merge commit log should have two parents")
+	//}
+	//differences, _, err := c.Diff(ctx, repository, commitLog.Parents[0], commitLog.Parents[1], -1, "")
```
Comment (nopcoder) [inline catalog/cataloger_merge_test.go:138]: can be removed

File: catalog/cataloger_diff_uncommitted_test.go [hunk 10]
```diff
@@ -5,11 +5,42 @@ import (
...
+	for i := 0; i < numOfFiles; i++ {
+		p := "/file" + strconv.Itoa(i)
+		testCatalogerCreateEntry(t, ctx, c, repository, "master", p, nil, "")
+		expectedDifferences = append(expectedDifferences, Difference{Type: DifferenceTypeAdded, Path: p})
+	}
+	const changesPerPage = 3
+	var differences Differences
+	var after string
+	for {
+		res, hasMore, err := c.DiffUncommitted(ctx, repository, "master", changesPerPage, after)
```
Comment (nopcoder) [inline catalog/cataloger_diff_uncommitted_test.go:28]: verify that `res` equal or less than `changesPerPage`

File: catalog/cataloger_list_commits_test.go [hunk 11]
```diff
@@ -357,9 +357,18 @@ func TestCataloger_ListCommits_LineageFromChild(t *testing.T) {
 	}
 	merge2, err := c.Merge(ctx, repository, "br_2_1", "master", "tester", "merge br_2_1 to master", nil)
 	testutil.MustDo(t, "merge br_2_1  into master", err)
-	if merge2.Differences[0].Type != DifferenceTypeChanged || merge2.Differences[0].Path != "master-file" {
-		t.Error("merge br_2_1 into master with unexpected results", merge2.Differences[0])
-	}
+	commitLog, err := c.GetCommit(ctx, repository, merge2.Reference)
+	testutil.MustDo(t, "get merge commit reference", err)
+	if len(commitLog.Parents) != 2 {
```
Comment (nopcoder) [inline catalog/cataloger_list_commits_test.go:362]: Also for all calls to Merge check Summary

**Comment by nopcoder** [comment]
Closes #575

**Review by johnnyaug — APPROVED** [review]

**Review by johnnyaug — APPROVED** [review]

**Pull request merged**